src/rgw/rgw_crypt.cc

```diff
@@ -0,0 +1,1314 @@
+    //drop info on log
+    if (*again) {
+      ldout(cct, 0) << "*again==true" << dendl;
+    }
+    return res;
+  }
+  off_t bl_ofs = 0;
+  if (cache.length() > 0) {
+      //append before operation.
+      off_t size = block_size - cache.length();
+      if (size > bl.length())
+        size = bl.length();
+      if (size > 0) {
+        char *src = bl.get_contiguous(0, size);
+        cache.append(src,size);
+        bl_ofs += size;
+      }
+      if (cache.length() == block_size) {
+        bufferlist data;
+        crypt->encrypt(cache, 0, block_size, data, ofs);
+        res = next->handle_data(data, ofs, phandle, pobj, again);
+        cache.clear();
+        ofs += block_size;
+        if (res != 0)
+          return res;
+      }
+    }
+  if (bl_ofs < bl.length()) {
+    off_t aligned_size = (bl.length() - bl_ofs) & ~(block_size - 1);
+    //save remainder
+    off_t remainder = (bl.length() - bl_ofs) - aligned_size;
+    if(remainder > 0) {
+      cache.append(bl.get_contiguous(bl_ofs + aligned_size, remainder), remainder);
+    }
+    if (aligned_size > 0) {
+      bufferlist data;
+      crypt->encrypt(bl, bl_ofs, aligned_size, data, ofs);
+      res=next->handle_data(data, ofs, phandle, pobj, again);
+      ofs += aligned_size;
+      if (res != 0)
+        return res;
+    }
+  }
+  if (bl.length() == 0) {
+    if (cache.length() > 0) {
+      /*flush cached data*/
+      bufferlist data;
+      crypt->encrypt(cache, 0, cache.length(), data, ofs);
+      res = next->handle_data(data, ofs, phandle, pobj, again);
+      ofs += cache.length();
+      cache.clear();
+      if (res != 0)
+        return res;
+    }
+    /*replicate 0-sized handle_data*/
+    res = next->handle_data(cache, ofs, phandle, pobj, again);
+  }
+  return res;
+}
+
+int RGWPutObj_BlockEncrypt::throttle_data(void *handle, const rgw_obj& obj, bool need_to_wait) {
+  return next->throttle_data(handle, obj, need_to_wait);
+}
+
+std::string create_random_key_selector() {
+  char random[AES_256_KEYSIZE];
+  if (get_random_bytes(&random[0], sizeof(random)) != 0) {
+    dout(0) << "ERROR: cannot get_random_bytes. " << dendl;
+    for (char& v:random) v=rand();
+  }
+  return std::string(random, sizeof(random));
+}
+
+int get_barbican_url(CephContext * const cct,
+                     std::string& url)
+{
+  url = cct->_conf->rgw_barbican_url;
+  if (url.empty()) {
+    ldout(cct, 0) << "ERROR: conf rgw_barbican_url is not set" << dendl;
+    return -EINVAL;
+  }
+
+  if (url[url.size() - 1] != '/') {
+    url.append("/");
+  }
+
+  return 0;
+}
+
+int request_key_from_barbican(CephContext *cct,
+                              boost::string_ref key_id,
+                              boost::string_ref key_selector,
+                              const std::string& barbican_token,
+                              std::string& actual_key) {
+  std::string secret_url;
+  if (get_barbican_url(cct, secret_url) < 0) {
+     return -EINVAL;
```

Comment:
if `get_barbican_url()` is returning an error code, we should return that instead of -EINVAL here